proposals/0002-testing-log-handler.md

```diff
@@ -0,0 +1,112 @@
+# Test Log Handler
+
+Authors: [Brian Maher](https://github.com/brimworks), [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+
+## Introduction
+
+Software operators rely on logging to provide insight into what is being done,
+therefore, testing log output is an important facet.
+
+## Motivation
+
+A variety of implementations have cropped up organically, but having a single
+test log handler to build off of will reduce the re-write churn.
+
+### Existing Implementations
+
+* [swift-log-testing](https://github.com/neallester/swift-log-testing)
+* [swift-distributed-actors/LogCapture](https://github.com/apple/swift-distributed-actors/blob/main/Sources/DistributedActorsTestKit/LogCapture.swift)
+* [swift-cluster-membership/LogCapture](https://github.com/apple/swift-cluster-membership/blob/main/Tests/SWIMTestKit/LogCapture.swift)
+* [RediStack/RedisLoggingTests](https://github.com/swift-server/RediStack/blob/master/Tests/RediStackIntegrationTests/RedisLoggingTests.swift)
+* [swift-log/TestLogger](https://github.com/apple/swift-log/blob/main/Tests/LoggingTests/TestLogger.swift#L52)
+
+
+## Features
+
+A testing log handler needs the following features:
+* The ability to capture log entries at the start of a unit test.
+* The ability to assert the captured log entries contain a matching message.
+* Work correctly in the face of parallel test execution.
+* Print log messages to the console if the log entry does not match a log entry
+  that is being asserted.
+* Bootstrap the test log handler as early as possible so we don't miss any
+  `Logger()` creations which may happen during setup.
+
+## Proposed Solution
+
+A global function to bootstrap the test log handler:
+
+```swift
+TestingLogHandler.bootstrap()
+```
+
+A way to create a new "container" which will capture all log messages that match
```

Comment:
I am not sure I follow the `container` principle yet. Should the testing log handler start by capturing all logs and metadata and then provide easy assertion APIs in the end? Potentially a connection to `swift-testing` that makes that as easy as possible would be great.

Reply:
> I propose only non-matching log messages will be printed to the console.

I don't think this is right, at least not by default; A "log capture" log handler should capture all logs, then we assert on those (or filter out and are able to eagerly drop some), but we should not cause prints to happen.

If an assertion fails, there should be a way to "there was an error (the error); dump all the logs captured". This is how I use logging of _vast_ quantities of logs, and I get the "important" logs dumped when an assertion has failed.

Reply:
> I don't think this is right, at least not by default; A "log capture" log handler should capture all logs, then we assert on those (or filter out and are able to eagerly drop some), but we should not cause prints to happen.

Let me tell you a story about when our team first attempted to add log assertions to our code base. Initially we added a dependency on [neallester/swift-log-testing](https://github.com/neallester/swift-log-testing) and we [followed the example](https://github.com/neallester/swift-log-testing/blob/master/Tests/SwiftLogTestingTests/ExampleTests.swift#L27-L34):

```swift
override func setUp() {
        TestLogMessages.bootstrap() 
}
```

We added in our log assertions and everything was great... until our CI pipeline broke and we realized that **none** of our log messages got printed, and since the logs were existing with signal 11 we had no idea what the problem was. In fact, this different design (print non-captured log messages) was partially motivated by this inoculating experience.

Therefore, I **highly** recommend that whatever final approach we land on will continue to default to printing logs at least when running tests that are not trying to use the log testing framework.

> If an assertion fails, there should be a way to "there was an error (the error); dump all the logs captured". This is how I use logging of vast quantities of logs, and I get the "important" logs dumped when an assertion has failed.

What if the code which performs the log assertion is never executed? I can think of many scenarios where I have experienced this: exception is thrown, process segfaults, etc.

> But I don't think "passed" execution should be logging anything by default

In the current design, if a log **matches** any container, then the message is not logged. Maybe I didn't make that part clear?

> in fact, tests which print much output are an anti pattern imho and stdout should be empty if nothing went wrong

Agreed... and if you use the test logging framework I propose, then any log message that matches an "expectation" won't be printed.

What do you think of this idea? Instead of having a container with a pre-filter as I currently have designed, what if I borrow some ideas from [java EasyMock](https://easymock.org/user-guide.html). More specifically, what if you could define a set of "expectations" for log messages, and then at the end of your test you can "verify" these expectations.

So, a typical usage might look like this:

```swift
class TestMyLogs: XCTestCase {
    override func setUp() {
        TestingLogHandler.bootstrap() 
    }
    func testExample() {
        // Logger may be created before the verifier:
        let logger = Logger(label: "Example")
        let verifier = LogVerifier()
        verifier.expect(label: "Example", level: .debug, message: Regex("^Message$"))
        logger.debug("Message")
        // Performs assertion(s) on all expectations. If any expectation fails, the expectation(s)
        // that fail are printed in the assertion failure. This call also resets all expectations.
        verifier.verify()
        // This call would fail, with an error indicating that there are no expectations set.
        verifier.verify()
    }
}
```

Reply:
Couldn't this behavior be a flag passed, so that people who need to debug/inspect their test executions more deeply can - and for those who don't, have their ideal clean stdout?

Reply:
> Couldn't this behavior be a flag passed, so that people who need to debug/inspect their test executions more deeply can - and for those who don't, have their ideal clean stdout?

In the current implementation I have proposed, just collect all the logs by specifying a pre-filter that selects all logs `{ true }`  and then you can get a "clean" stdout during that test run, and you can iterate over the container of logs collected to do all the normal assertion stuff you might want to do. I don't advocate for using it in that way, since I find it more useful to be selective up-front using a pre-filter, but the current implementation I have suggested supports both uses.

Reply:
Similar to my comment above. I am not convinced we need this complicated container API if we should expect users to properly accept `Logger`s at the API level. We can then just store all log messages in the test log handler and assert on that. This should also work during parallel test execution